numpy/core/shape_base.py

```diff
@@ -204,11 +206,22 @@ def atleast_3d(*arys):
         return res
 
 
-def _vstack_dispatcher(tup):
-    return tup
+def _arrays_for_stack_dispatcher(arrays, stacklevel=4):
+    if not hasattr(arrays, '__getitem__') and hasattr(arrays, '__iter__'):
+        warnings.warn('arrays to stack must be passed as a "sequence" type '
+                      'such as list or tuple. Support for non-sequence '
+                      'iterables such as generators is deprecated as of '
+                      'NumPy 1.16 and will raise an error in the future.',
+                      FutureWarning, stacklevel=stacklevel)
```

Comment:
This raises an interesting point - is wrapping everything in `__array_ufunc__` going to make all our warnings be emitted within the `__array_ufunc__` dispatcher?

Reply:
In this case, yes.

In general: warnings about changed functions arguments/signatures should go in the dispatcher, since they are relevant whether you are using NumPy or not. Warnings about changed function behavior should remain in the implementations.